### PR TITLE
Adding the Xavier Initializer

### PR DESCRIPTION
**Pull request by abhinavarora**

### CODE REVIEW

File: python/paddle/v2/framework/initializer.py [hunk 1]
```diff
@@ -20,6 +24,41 @@ def __call__(self, param, block):
...
+            fan_in = fan_out = 1
+        elif len(shape) == 1:
+            fan_in = fan_out = shape[0]
+        elif len(shape) == 2:
+            # This is the case for simple matrix multiply
+            fan_in = shape[0]
+            fan_out = shape[1]
+        else:
+            # Assume this to be a convolutional kernel
+            # In Paddlepaddle, the shape of the kernel is like:
```
Comment (pengli09) [inline python/paddle/v2/framework/initializer.py]: s/Paddlepaddle/PaddlePaddle/g

File: python/paddle/v2/framework/initializer.py [hunk 2]
```diff
@@ -156,3 +195,93 @@ def __call__(self, var, block):
...
+    is sqrt(2/ (fan_in + fan_out)).
+
+    References:
+        [1] Understanding the difficulty of training deep feedforward neural
+            networks. International conference on artificial intelligence and
+            statistics.
+            (http://proceedings.mlr.press/v9/glorot10a.html)
+    """
+
+    def __init__(self, uniform=True, fan_in=None, fan_out=None, seed=0):
```
Comment (pengli09) [inline python/paddle/v2/framework/initializer.py]: Normal distribution is used more frequently than Uniform distribution. Shall we change the default value of `uniform` to `False`

File: python/paddle/v2/framework/initializer.py [hunk 3]
```diff
@@ -156,3 +195,93 @@ def __call__(self, var, block):
...
+        """Constructor for XavierInitializer
+
+        Args:
```
Comment (pengli09) [inline python/paddle/v2/framework/initializer.py]: Please add descriptions for the default values.

File: python/paddle/v2/framework/initializer.py [hunk 4]
```diff
@@ -156,3 +195,93 @@ def __call__(self, var, block):
...
+
+        Returns:
+            the initialization op
+        """
+        assert isinstance(var, framework.Variable)
+        assert isinstance(block, framework.Block)
+        f_in, f_out = self._compute_fans(var)
+
+        # If fan_in and fan_out are passed, use them
+        fan_in = self._fan_in or f_in
```
Comment (pengli09) [inline python/paddle/v2/framework/initializer.py]: `f_in` will be used if `self._fan_in` is 0, which is counter-intuitive. Therefore I think `f_in if self._fan_in is None else self._fan_in`. 0 is not a proper value for `self._fan_in`, which means, when encountering a 0, either the user is using it intentionally for debugging purpose or an error occurs. Therefore,  I think failing deterministically will be a better choice.

File: python/paddle/v2/framework/initializer.py [hunk 5]
```diff
@@ -156,3 +195,93 @@ def __call__(self, var, block):
...
+        fan_out = self._fan_out or f_out
```
Comment (pengli09) [inline python/paddle/v2/framework/initializer.py]: see above

File: python/paddle/v2/framework/initializer.py [hunk 6]
```diff
@@ -156,3 +195,93 @@ def __call__(self, var, block):
...
+
+        if self._uniform:
+            limit = np.sqrt(6.0 / float(fan_in + fan_out))
+            op = block.prepend_op(
+                type="uniform_random",
+                outputs={"Out": var},
+                attrs={
+                    "shape": var.shape,
+                    "data_type": int(var.data_type),
```
Comment (pengli09) [inline python/paddle/v2/framework/initializer.py]: I'm not familiar with Block. Is `int` correct?

File: python/paddle/v2/framework/tests/test_initializer.py [hunk 7]
```diff
@@ -116,5 +117,112 @@ def test_normal_initializer(self):
...
+            lod_level=0,
+            name="param",
+            initializer=initializer.XavierInitializer())
+        self.assertEqual(len(block.ops), 1)
+        init_op = block.ops[0]
+        self.assertEqual(init_op.type, 'uniform_random')
+        limit = np.sqrt(6.0 / (param.shape[0] + param.shape[1]))
+        self.assertAlmostEqual(init_op.attr('min'), -limit, delta=DELTA)
+        self.assertAlmostEqual(init_op.attr('max'), limit, delta=DELTA)
+        self.assertEqual(init_op.attr('seed'), 0)
```
Comment (pengli09) [inline python/paddle/v2/framework/tests/test_initializer.py]: I think we should also test whether seed can be set properly.

File: python/paddle/v2/framework/initializer.py [hunk 8]
```diff
@@ -156,3 +195,93 @@ def __call__(self, var, block):
...
+
+    This class implements the Xavier weight initializer from the paper
+    Understanding the difficulty of training deep feedforward neural
+    networks[1] by Xavier Glorot and Yoshua Bengio.
+
+    This initializer is designed to keep the scale of the gradients
+    approximately same in all the layers. In case of Uniform distribution,
+    the range is [-x, x], where x = sqrt(6 / (fan_in + fan_out)).
+    In case of Normal distribution, the mean is 0 and the standard deviation
+    is sqrt(2/ (fan_in + fan_out)).
```
Comment (pengli09) [inline python/paddle/v2/framework/initializer.py]: The recommended scale/std for different nonlinearity functions are different. I think we can borrow the idea from PyTorch to make this initializer more general.
https://github.com/pytorch/pytorch/blob/master/torch/nn/init.py#L8
https://github.com/pytorch/pytorch/blob/master/torch/nn/init.py#L184

File: python/paddle/v2/framework/initializer.py [hunk 9]
```diff
@@ -156,3 +195,93 @@ def __call__(self, var, block):
...
+
+    References:
+        [1] Understanding the difficulty of training deep feedforward neural
+            networks. International conference on artificial intelligence and
+            statistics.
+            (http://proceedings.mlr.press/v9/glorot10a.html)
+    """
```
Comment (lcy-seso) [inline python/paddle/v2/framework/initializer.py]: I like to read your comments~ Quite easy to follow.

Just for my personal interests. I remember this is a recommended initializer for tanh, and usually, the usually recommended initializer varies according to different activations.

I think we can just add all these helpful initializers into our framework for the first goal, and later to choose a much better initialization strategy to encode the best practice.

File: python/paddle/v2/framework/initializer.py (hunk 2 above)
Comment (lcy-seso) [reply python/paddle/v2/framework/initializer.py]: I think uniform distribution is acceptable for me, my concern is we are not in a stage that we are intensively encoding the best practices from the community into our framework. We can carefully tune this later.

But uniform distribution from [-1, 1] seems terrible. At least, for example, we can reduce the region limit into le-3.

Comment (abhinavarora) [reply python/paddle/v2/framework/initializer.py]: I was looking at tensorflow where they keep uniform as the default. That is why I chose uniform as the default.

File: python/paddle/v2/framework/initializer.py (hunk 6 above)
Comment (abhinavarora) [reply python/paddle/v2/framework/initializer.py]: Yes this is correct. We pass the data type as integer which is then mapped to a given type.

File: python/paddle/v2/framework/initializer.py (hunk 8 above)
Comment (abhinavarora) [reply python/paddle/v2/framework/initializer.py]: The original paper does not talk about Relus and gain because this paper was before the time when RELUs became popular. I also looked at tensorflow and keras source code. They have kept this initialization as it is defined in the paper. Maybe we can merge this for now and then add the gain later in a separate PR when we have more knowledge about it. I can read few more papers to see how the gain attribute is used. I do not think it is right to borrow the idea directly without researching it thoroughly. We can merge this and can look at this in more detail after refactoring is complete. What do you suggest?

File: python/paddle/v2/framework/initializer.py (hunk 3 above)
Comment (abhinavarora) [reply python/paddle/v2/framework/initializer.py]: Thank you for the feedback. Could you give me an example of the description you are referring to. I have talked about fan_in and fan_out in the `Args` section of my docstring.

File: python/paddle/v2/framework/initializer.py (hunk 1 above)
Comment (abhinavarora) [reply python/paddle/v2/framework/initializer.py]: Fixed in 4e1fa1b305a9d3662950eb69b5cbb1c28bc166f8

File: python/paddle/v2/framework/initializer.py (hunk 4 above)
Comment (abhinavarora) [reply python/paddle/v2/framework/initializer.py]: Fixed in 4e1fa1b305a9d3662950eb69b5cbb1c28bc166f8

File: python/paddle/v2/framework/tests/test_initializer.py (hunk 7 above)
Comment (abhinavarora) [reply python/paddle/v2/framework/tests/test_initializer.py]: Fixed in 4e1fa1b305a9d3662950eb69b5cbb1c28bc166f8

File: python/paddle/v2/framework/initializer.py (hunk 9 above)
Comment (abhinavarora) [reply python/paddle/v2/framework/initializer.py]: Thank you for the feedback. I agree with you.

File: python/paddle/v2/framework/initializer.py (hunk 2 above)
Comment (lcy-seso) [reply python/paddle/v2/framework/initializer.py]: Got it~ thank you for the information.

**Comment by lcy-seso** [comment]
Not changing the default uniform distribution is acceptable for me.  Because for me this is not a top concern currently and I do not have a strong evidence to support any reasonable change.

File: python/paddle/v2/framework/initializer.py (hunk 3 above)
Comment (pengli09) [reply python/paddle/v2/framework/initializer.py]: For example, "uniform: .... Default value is True."
